packages/adapter-utils/src/utils.ts

```diff
@@ -101,33 +125,112 @@ const fieldMapperGenerator = (
   )
 }
 
-export type TransformFuncArgs = {
-  value: Value
-  path?: ElemID
-  field?: Field
+const removeEmptyParts = (value: Value, allowEmpty: boolean): Value => {
+  if (Array.isArray(value)) {
+    const filtered = value.filter(isDefined)
+    return filtered.length === 0 && (value.length > 0 || !allowEmpty) ? undefined : filtered
+  }
+  if (_.isPlainObject(value)) {
+    const filtered = _.omitBy(value, _.isUndefined)
+    return _.isEmpty(filtered) && (!_.isEmpty(value) || !allowEmpty) ? undefined : filtered
+  }
+  return value
 }
-export type TransformFunc = (args: TransformFuncArgs) => Promise<Value> | Value | undefined
 
-export type TransformFuncSync = (args: TransformFuncArgs) => lowerDashTypes.NonPromise<Value> | undefined
+const recurseIntoValue = (
+  newVal: Value,
+  transformFunc: (value: Value, keyPathID?: ElemID, field?: Field) => Value,
+  strict: boolean,
+  allowEmpty: boolean,
+  isAsync: boolean,
+  keyPathID?: ElemID,
+  field?: Field,
+  fieldType?: TypeElement,
+): Value => {
+  if (newVal === undefined) {
+    return undefined
+  }
 
-type TransformValuesBaseArgs = {
-  values: Value
-  type: ObjectType | TypeMap | MapType | ListType
-  strict?: boolean
-  pathID?: ElemID
-  isTopLevel?: boolean
-  allowEmpty?: boolean
+  if (isReferenceExpression(newVal)) {
+    return newVal
+  }
+
+  const listMapFunc = isAsync
+    ? (list: Iterable<Value>,
+      mapFunc: (t: Value, index: number) => Promise<unknown>) => toArrayAsync(mapAsync(list, mapFunc))
+    : _.map
+  const objMapFunc = isAsync
+    ? (obj: Iterable<Value>, mapFunc: (val: Value, key: string) => Promise<unknown>) => mapValuesAsync(obj, mapFunc)
+    : _.mapValues
+
+  if (field && isListType(fieldType)) {
+    const transformListInnerValue = (item: Value, index?: number): Value => (
+      transformFunc(
+        item,
+        !_.isUndefined(index) ? keyPathID?.createNestedID(String(index)) : keyPathID,
+        new Field(
+          field.parent,
+          field.name,
+          fieldType.refInnerType,
+          field.annotations
+        ),
+      )
+    )
+    if (!_.isArray(newVal)) {
+      if (strict) {
+        log.debug(`Array value and isListType mis-match for field - ${field.name}. Got non-array for ListType.`)
+      }
+      return transformListInnerValue(newVal)
+    }
+    return listMapFunc(newVal, (v: Value, i: number) => transformListInnerValue(v, i))
+  }
+  if (_.isArray(newVal)) {
+    // Even fields that are not defined as ListType can have array values
+    return listMapFunc(
+      newVal, (item: Value, index: number) => transformFunc(item, keyPathID?.createNestedID(String(index)), field)
```

Comment:
1. types and not needed in the callback, they can be inferred from the `listMapFunc`
2. we should be consistent with the argument names
```suggestion
    return listMapFunc(newVal, (item, index) => transformListInnerValue(item, index))
  }
  if (_.isArray(newVal)) {
    // Even fields that are not defined as ListType can have array values
    return listMapFunc(
      newVal, (item, index) => transformFunc(item, keyPathID?.createNestedID(String(index)), field)
```

Reply:
I'm getting errors when removing the types :(
```
../adapter-utils/src/utils.ts:185:33 - error TS7006: Parameter 'item' implicitly has an 'any' type.

185     return listMapFunc(newVal, (item, index) => transformListInnerValue(item, index))
                                    ~~~~

../adapter-utils/src/utils.ts:185:39 - error TS7006: Parameter 'index' implicitly has an 'any' type.

185     return listMapFunc(newVal, (item, index) => transformListInnerValue(item, index))
```